### PR TITLE
Typescriptify

### PR DESCRIPTION
**Pull request by rejas**

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -1,70 +1,85 @@
...
-  // React 17 changed where it attaches it's event listeners.
+  // React 17 changed where it attaches its event listeners.
   // React 16 attached all event listeners to document. React 17 attaches all event
-  // listeners to the react root node.
+  // listeners to the React root node.
   // Given an event handler of type 'click' attaches a global click handler onto
   // document, the click event would bubble up to the newly attached document click
-  // handler (which is, in most cases, an unexpected bahvior). We can prevent this
-  // from attaching specific events to the react root node instead.
+  // handler (which is, in most cases, an unexpected behavior). We can prevent this
```
Comment (peterhass) [inline src/index.ts]: image

File: tsconfig.json [hunk 2]
```diff
@@ -0,0 +1,29 @@
+{
+  "compilerOptions": {
+    "target": "es2016",
+    "module": "esnext",
+    "allowJs": true,
```
Comment (peterhass) [inline tsconfig.json:5]: false? js is not needed anymore i guess

File: src/index.ts [hunk 3]
```diff
@@ -1,70 +1,85 @@
 import debounce from 'lodash/debounce';
 import throttle from 'lodash/throttle';
 
+interface EventData {
```
Comment (peterhass) [inline src/index.ts]: Would rather use `type` here. You only benefit from Interfaces for global types that might be extended like `window`. Whenever you're sure that you're working with a non-extendable type, go for `type`. (Pretty much: Always use `type` until you can't 😀)

File: src/index.ts [hunk 4]
```diff
@@ -100,14 +119,7 @@ const removeListener = (event, id) => {
...
 
-/**
- * @function
- * @param {string} event
- * @param {function} callback
- * @return {removeEventListener|noop}
- */
-const addListener = (event, callback) => {
-  if (typeof event !== 'string') throw new Error('Event name required');
+const addListener = (event:string, callback: HandlerFunction) => {
```
Comment (peterhass) [inline src/index.ts]: we could mirror the type of `document.addEventListener` here:     
```ts
addEventListener<K extends keyof DocumentEventMap>(type: K, listener: (this: Document, ev: DocumentEventMap[K]) => any, options?: boolean | AddEventListenerOptions): void;
```

File: src/index.ts (hunk 1 above)
Comment (rejas) [reply src/index.ts]: ???

Comment (peterhass) [reply src/index.ts]: Behavior -> behaviour

File: .eslintrc [hunk 5]
```diff
@@ -1,9 +1,44 @@
...
+      ],
+      "env": {
+        "mocha": true
+      },
+      "rules": {
+        "no-unused-expressions": "off"
+      },
+      "extends": [
+        "nebenan-base",
+        "airbnb-typescript/base"
```
Comment (iamuchejude) [inline .eslintrc]: We have an eslint preset for TypeScript

File: src/index.ts [hunk 6]
```diff
@@ -1,70 +1,84 @@
...
+
+type EventData = {
+  listeners: Record<string, HandlerFunction>;
+  lastIndex: number;
+  listenersLength: number;
+};
+
+type EventSettings = {
+  emitter?: Document | typeof globalThis;
+  wrapper?: (callback: Function) => Function;
```
Comment (iamuchejude) [inline src/index.ts]: We should avoid anything similar to `any`. `Function` is basically `any` function.

It's practically to tell the structure of `Function` as it's generic, so type-safety is incomplete.

A fix would be to define types to explicitly describe your function.

File: src/index.ts [hunk 7]
```diff
@@ -1,70 +1,84 @@
...
-const defaultSettings = {};
-
+const defaultSettings: EventSettings = {};
 
 const createEventSettings = () => {
   settingsMap.resize = {
-    emitter: global,
-    wrapper(callback) { return debounce(callback, RESIZE_RATE); },
+    emitter: globalThis,
+    wrapper(callback) { return debounce(callback as any, RESIZE_RATE); },
```
Comment (iamuchejude) [inline src/index.ts]: This won't be necessary if the function type is properly described and the other `anys`

**Review by iamuchejude — CHANGES_REQUESTED** [review]
Getting there 🛥️

**Pull request closed without merging**